File: Utils/Cpp_Samples/bin_create.cpp
///////////////////////////////////////////////////////////////////////////////
// FuryUtils Sample
//
// Creating and saving a Bin game level file
//
// using:
//
//    Bin
//        class for reading / writing Bin game level files.
//    Exceptions::Exception
//       exception class for reporting errors.
//
//
//   Suggested inputs from the testassets: N/A
//
///////////////////////////////////////////////////////////////////////////////

#include <ios>
#include <fstream>
#include "../include/FuryUtils.hpp"

void bin_create(const char * fileName1) {
	printf("bin_create Sample\n");
	
	try {

		// create a Bin archive
		FuryUtils::Archive::Bin bin;
		bin.SetComment("This is a test comment");
		bin.mapWidth = 30;
		bin.mapHeight = 25;

		// Get a buffer for the completed file and save it to a file
		std::ofstream outFile(fileName1, std::ios::binary | std::ios::trunc);
		std::vector<uint8_t> buffer;
		bin.Convert(buffer, FuryUtils::Archive::Bin::Compressed);
		outFile.write((char *)(buffer.data()), buffer.size());
		
		return ;
	}
	catch (FuryUtils::Exceptions::Exception e)
	{
		printf("Error:\n\n%d %s\n", e._errorCode, e._errorString.c_str());
		return;
	}
	return;
}

File: CS_Samples/bin_create.cs
///////////////////////////////////////////////////////////////////////////////
///
/// FuryUtils Sample
/// 
/// Creating and saving a Bin game level file
/// 
/// using:
/// 
///    Bin
///         class for handling Bin game level files.
///    FuryException
///         exception class for reporting errors.
///
///
/// Suggested inputs from the testassets: N/A
/// 
///////////////////////////////////////////////////////////////////////////////

using carbon14.FuryStudio.Utils;

namespace carbon14.FuryStudio.CS_Samples
{
    internal class bin_create
    {
        public static void Execute(string fileName1)
        {
            Console.WriteLine("bin_create Sample");

            try
            {
                // create a new Bin file
                using (Bin bin = new Bin())
                {
                    // set properties
                    bin.data.mapWidth = 30;
                    bin.data.mapHeight = 25;
                    bin.Comment = "This is a test comment";

                    // Get a buffer to save
                    byte[]? buffer = bin.Convert(Bin.ConversionType.Compressed);

                    // Save the buffer to a file.
                    if (buffer != null)
                    {
                        using (FileStream fs = new FileStream(fileName1, FileMode.Create))
                        {

                            fs.Write(buffer, 0, buffer.Length);
                        }
                    }
                }
            }
            catch (FuryException ex)
            {
                Console.WriteLine($"Error: {ex.ErrorCode} {ex.Message}");
            }
        }

    }
}
